// medusa-interfaces/dist/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SearchService = exports.OauthService = exports.NotificationService = exports.FileService = exports.FulfillmentService = exports.PaymentService = exports.BaseService = void 0;
var base_service_1 = require("./base-service");
Object.defineProperty(exports, "BaseService", { enumerable: true, get: function () { return __importDefault(base_service_1).default; } });
var payment_service_1 = require("./payment-service");
Object.defineProperty(exports, "PaymentService", { enumerable: true, get: function () { return __importDefault(payment_service_1).default; } });
var fulfillment_service_1 = require("./fulfillment-service");
Object.defineProperty(exports, "FulfillmentService", { enumerable: true, get: function () { return __importDefault(fulfillment_service_1).default; } });
var file_service_1 = require("./file-service");
Object.defineProperty(exports, "FileService", { enumerable: true, get: function () { return __importDefault(file_service_1).default; } });
var notification_service_1 = require("./notification-service");
Object.defineProperty(exports, "NotificationService", { enumerable: true, get: function () { return __importDefault(notification_service_1).default; } });
var oauth_service_1 = require("./oauth-service");
Object.defineProperty(exports, "OauthService", { enumerable: true, get: function () { return __importDefault(oauth_service_1).default; } });
var search_service_1 = require("./search-service");
Object.defineProperty(exports, "SearchService", { enumerable: true, get: function () { return __importDefault(search_service_1).default; } });
