// medusa-payment-stripe/services/stripe-provider.js
"use strict";

function _instanceof(left, right) { if (right != null && typeof Symbol !== "undefined" && right[Symbol.hasInstance]) { return !!right[Symbol.hasInstance](left); } else { return left instanceof right; } }

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _stripe = _interopRequireDefault(require("stripe"));

var _medusaInterfaces = require("medusa-interfaces");

var _dist = require("@medusajs/medusa/dist");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) { symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); } keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(_next, _throw); } }

function _asyncToGenerator(fn) { return function () { var self = this, args = arguments; return new Promise(function (resolve, reject) { var gen = fn.apply(self, args); function _next(value) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value); } function _throw(err) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err); } _next(undefined); }); }; }

function _classCallCheck(instance, Constructor) { if (!_instanceof(instance, Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } else if (call !== void 0) { throw new TypeError("Derived constructors may only return object or undefined"); } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var StripeProviderService = /*#__PURE__*/function (_PaymentService) {
  _inherits(StripeProviderService, _PaymentService);

  var _super = _createSuper(StripeProviderService);

  function StripeProviderService(_ref, options) {
    var _this;

    var customerService = _ref.customerService,
        totalsService = _ref.totalsService,
        regionService = _ref.regionService;

    _classCallCheck(this, StripeProviderService);

    _this = _super.call(this);
    /**
     * Required Stripe options:
     *  {
     *    api_key: "stripe_secret_key", REQUIRED
     *    webhook_secret: "stripe_webhook_secret", REQUIRED
     *    // Use this flag to capture payment immediately (default is false)
     *    capture: true
     *  }
     */

    _this.options_ = options;
    /** @private @const {Stripe} */

    _this.stripe_ = (0, _stripe["default"])(options.api_key);
    /** @private @const {CustomerService} */

    _this.customerService_ = customerService;
    /** @private @const {RegionService} */

    _this.regionService_ = regionService;
    /** @private @const {TotalsService} */

    _this.totalsService_ = totalsService;
    return _this;
  }
  /**
   * Fetches Stripe payment intent. Check its status and returns the
   * corresponding Medusa status.
   * @param {object} paymentData - payment method data from cart
   * @returns {string} the status of the payment intent
   */


  _createClass(StripeProviderService, [{
    key: "getStatus",
    value: function () {
      var _getStatus = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee(paymentData) {
        var id, paymentIntent;
        return regeneratorRuntime.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                id = paymentData.id;
                _context.next = 3;
                return this.stripe_.paymentIntents.retrieve(id);

              case 3:
                paymentIntent = _context.sent;
                _context.t0 = paymentIntent.status;
                _context.next = _context.t0 === "requires_payment_method" ? 7 : _context.t0 === "requires_confirmation" ? 7 : _context.t0 === "processing" ? 7 : _context.t0 === "requires_action" ? 8 : _context.t0 === "canceled" ? 9 : _context.t0 === "requires_capture" ? 10 : _context.t0 === "succeeded" ? 10 : 11;
                break;

              case 7:
                return _context.abrupt("return", _dist.PaymentSessionStatus.PENDING);

              case 8:
                return _context.abrupt("return", _dist.PaymentSessionStatus.REQUIRES_MORE);

              case 9:
                return _context.abrupt("return", _dist.PaymentSessionStatus.CANCELED);

              case 10:
                return _context.abrupt("return", _dist.PaymentSessionStatus.AUTHORIZED);

              case 11:
                return _context.abrupt("return", _dist.PaymentSessionStatus.PENDING);

              case 12:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      function getStatus(_x) {
        return _getStatus.apply(this, arguments);
      }

      return getStatus;
    }()
    /**
     * Fetches a customers saved payment methods if registered in Stripe.
     * @param {object} customer - customer to fetch saved cards for
     * @returns {Promise<Array<object>>} saved payments methods
     */

  }, {
    key: "retrieveSavedMethods",
    value: function () {
      var _retrieveSavedMethods = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee2(customer) {
        var methods;
        return regeneratorRuntime.wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                if (!(customer.metadata && customer.metadata.stripe_id)) {
                  _context2.next = 5;
                  break;
                }

                _context2.next = 3;
                return this.stripe_.paymentMethods.list({
                  customer: customer.metadata.stripe_id,
                  type: "card"
                });

              case 3:
                methods = _context2.sent;
                return _context2.abrupt("return", methods.data);

              case 5:
                return _context2.abrupt("return", Promise.resolve([]));

              case 6:
              case "end":
                return _context2.stop();
            }
          }
        }, _callee2, this);
      }));

      function retrieveSavedMethods(_x2) {
        return _retrieveSavedMethods.apply(this, arguments);
      }

      return retrieveSavedMethods;
    }()
    /**
     * Fetches a Stripe customer
     * @param {string} customerId - Stripe customer id
     * @returns {Promise<object>} Stripe customer
     */

  }, {
    key: "retrieveCustomer",
    value: function () {
      var _retrieveCustomer = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee3(customerId) {
        return regeneratorRuntime.wrap(function _callee3$(_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                if (customerId) {
                  _context3.next = 2;
                  break;
                }

                return _context3.abrupt("return", Promise.resolve());

              case 2:
                return _context3.abrupt("return", this.stripe_.customers.retrieve(customerId));

              case 3:
              case "end":
                return _context3.stop();
            }
          }
        }, _callee3, this);
      }));

      function retrieveCustomer(_x3) {
        return _retrieveCustomer.apply(this, arguments);
      }

      return retrieveCustomer;
    }()
    /**
     * Creates a Stripe customer using a Medusa customer.
     * @param {object} customer - Customer data from Medusa
     * @returns {Promise<object>} Stripe customer
     */

  }, {
    key: "createCustomer",
    value: function () {
      var _createCustomer = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee4(customer) {
        var stripeCustomer;
        return regeneratorRuntime.wrap(function _callee4$(_context4) {
          while (1) {
            switch (_context4.prev = _context4.next) {
              case 0:
                _context4.prev = 0;
                _context4.next = 3;
                return this.stripe_.customers.create({
                  email: customer.email
                });

              case 3:
                stripeCustomer = _context4.sent;

                if (!customer.id) {
                  _context4.next = 7;
                  break;
                }

                _context4.next = 7;
                return this.customerService_.update(customer.id, {
                  metadata: {
                    stripe_id: stripeCustomer.id
                  }
                });

              case 7:
                return _context4.abrupt("return", stripeCustomer);

              case 10:
                _context4.prev = 10;
                _context4.t0 = _context4["catch"](0);
                throw _context4.t0;

              case 13:
              case "end":
                return _context4.stop();
            }
          }
        }, _callee4, this, [[0, 10]]);
      }));

      function createCustomer(_x4) {
        return _createCustomer.apply(this, arguments);
      }

      return createCustomer;
    }()
    /**
     * Creates a Stripe payment intent.
     * If customer is not registered in Stripe, we do so.
     * @param {object} cart - cart to create a payment for
     * @returns {object} Stripe payment intent
     */

  }, {
    key: "createPayment",
    value: function () {
      var _createPayment = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee5(cart) {
        var _cart$context$payment, _cart$context, _this$options_;

        var customer_id, region_id, email, _yield$this$regionSer, currency_code, amount, intentRequest, _customer$metadata, customer, stripeCustomer, _stripeCustomer;

        return regeneratorRuntime.wrap(function _callee5$(_context5) {
          while (1) {
            switch (_context5.prev = _context5.next) {
              case 0:
                customer_id = cart.customer_id, region_id = cart.region_id, email = cart.email;
                _context5.next = 3;
                return this.regionService_.retrieve(region_id);

              case 3:
                _yield$this$regionSer = _context5.sent;
                currency_code = _yield$this$regionSer.currency_code;
                _context5.next = 7;
                return this.totalsService_.getTotal(cart);

              case 7:
                amount = _context5.sent;
                intentRequest = {
                  description: (_cart$context$payment = cart === null || cart === void 0 ? void 0 : (_cart$context = cart.context) === null || _cart$context === void 0 ? void 0 : _cart$context.payment_description) !== null && _cart$context$payment !== void 0 ? _cart$context$payment : (_this$options_ = this.options_) === null || _this$options_ === void 0 ? void 0 : _this$options_.payment_description,
                  amount: Math.round(amount),
                  currency: currency_code,
                  setup_future_usage: "on_session",
                  capture_method: this.options_.capture ? "automatic" : "manual",
                  metadata: {
                    cart_id: "".concat(cart.id)
                  }
                };

                if (!customer_id) {
                  _context5.next = 23;
                  break;
                }

                _context5.next = 12;
                return this.customerService_.retrieve(customer_id);

              case 12:
                customer = _context5.sent;

                if (!((_customer$metadata = customer.metadata) !== null && _customer$metadata !== void 0 && _customer$metadata.stripe_id)) {
                  _context5.next = 17;
                  break;
                }

                intentRequest.customer = customer.metadata.stripe_id;
                _context5.next = 21;
                break;

              case 17:
                _context5.next = 19;
                return this.createCustomer({
                  email: email,
                  id: customer_id
                });

              case 19:
                stripeCustomer = _context5.sent;
                intentRequest.customer = stripeCustomer.id;

              case 21:
                _context5.next = 27;
                break;

              case 23:
                _context5.next = 25;
                return this.createCustomer({
                  email: email
                });

              case 25:
                _stripeCustomer = _context5.sent;
                intentRequest.customer = _stripeCustomer.id;

              case 27:
                _context5.next = 29;
                return this.stripe_.paymentIntents.create(intentRequest);

              case 29:
                return _context5.abrupt("return", _context5.sent);

              case 30:
              case "end":
                return _context5.stop();
            }
          }
        }, _callee5, this);
      }));

      function createPayment(_x5) {
        return _createPayment.apply(this, arguments);
      }

      return createPayment;
    }()
    /**
     * Retrieves Stripe payment intent.
     * @param {object} data - the data of the payment to retrieve
     * @returns {Promise<object>} Stripe payment intent
     */

  }, {
    key: "retrievePayment",
    value: function () {
      var _retrievePayment = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee6(data) {
        return regeneratorRuntime.wrap(function _callee6$(_context6) {
          while (1) {
            switch (_context6.prev = _context6.next) {
              case 0:
                _context6.prev = 0;
                return _context6.abrupt("return", this.stripe_.paymentIntents.retrieve(data.id));

              case 4:
                _context6.prev = 4;
                _context6.t0 = _context6["catch"](0);
                throw _context6.t0;

              case 7:
              case "end":
                return _context6.stop();
            }
          }
        }, _callee6, this, [[0, 4]]);
      }));

      function retrievePayment(_x6) {
        return _retrievePayment.apply(this, arguments);
      }

      return retrievePayment;
    }()
    /**
     * Gets a Stripe payment intent and returns it.
     * @param {object} sessionData - the data of the payment to retrieve
     * @returns {Promise<object>} Stripe payment intent
     */

  }, {
    key: "getPaymentData",
    value: function () {
      var _getPaymentData = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee7(sessionData) {
        return regeneratorRuntime.wrap(function _callee7$(_context7) {
          while (1) {
            switch (_context7.prev = _context7.next) {
              case 0:
                _context7.prev = 0;
                return _context7.abrupt("return", this.stripe_.paymentIntents.retrieve(sessionData.data.id));

              case 4:
                _context7.prev = 4;
                _context7.t0 = _context7["catch"](0);
                throw _context7.t0;

              case 7:
              case "end":
                return _context7.stop();
            }
          }
        }, _callee7, this, [[0, 4]]);
      }));

      function getPaymentData(_x7) {
        return _getPaymentData.apply(this, arguments);
      }

      return getPaymentData;
    }()
    /**
     * Authorizes Stripe payment intent by simply returning
     * the status for the payment intent in use.
     * @param {object} sessionData - payment session data
     * @param {object} context - properties relevant to current context
     * @returns {Promise<{ status: string, data: object }>} result with data and status
     */

  }, {
    key: "authorizePayment",
    value: function () {
      var _authorizePayment = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee8(sessionData) {
        var context,
            stat,
            _args8 = arguments;
        return regeneratorRuntime.wrap(function _callee8$(_context8) {
          while (1) {
            switch (_context8.prev = _context8.next) {
              case 0:
                context = _args8.length > 1 && _args8[1] !== undefined ? _args8[1] : {};
                _context8.next = 3;
                return this.getStatus(sessionData.data);

              case 3:
                stat = _context8.sent;
                _context8.prev = 4;
                return _context8.abrupt("return", {
                  data: sessionData.data,
                  status: stat
                });

              case 8:
                _context8.prev = 8;
                _context8.t0 = _context8["catch"](4);
                throw _context8.t0;

              case 11:
              case "end":
                return _context8.stop();
            }
          }
        }, _callee8, this, [[4, 8]]);
      }));

      function authorizePayment(_x8) {
        return _authorizePayment.apply(this, arguments);
      }

      return authorizePayment;
    }()
  }, {
    key: "updatePaymentData",
    value: function () {
      var _updatePaymentData = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee9(sessionData, update) {
        return regeneratorRuntime.wrap(function _callee9$(_context9) {
          while (1) {
            switch (_context9.prev = _context9.next) {
              case 0:
                _context9.prev = 0;
                return _context9.abrupt("return", this.stripe_.paymentIntents.update(sessionData.id, _objectSpread({}, update.data)));

              case 4:
                _context9.prev = 4;
                _context9.t0 = _context9["catch"](0);
                throw _context9.t0;

              case 7:
              case "end":
                return _context9.stop();
            }
          }
        }, _callee9, this, [[0, 4]]);
      }));

      function updatePaymentData(_x9, _x10) {
        return _updatePaymentData.apply(this, arguments);
      }

      return updatePaymentData;
    }()
    /**
     * Updates Stripe payment intent.
     * @param {object} sessionData - payment session data.
     * @param {object} update - objec to update intent with
     * @returns {object} Stripe payment intent
     */

  }, {
    key: "updatePayment",
    value: function () {
      var _updatePayment = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee10(sessionData, cart) {
        var _cart$customer, _cart$customer$metada, stripeId;

        return regeneratorRuntime.wrap(function _callee10$(_context10) {
          while (1) {
            switch (_context10.prev = _context10.next) {
              case 0:
                _context10.prev = 0;
                stripeId = ((_cart$customer = cart.customer) === null || _cart$customer === void 0 ? void 0 : (_cart$customer$metada = _cart$customer.metadata) === null || _cart$customer$metada === void 0 ? void 0 : _cart$customer$metada.stripe_id) || undefined;

                if (!(stripeId !== sessionData.customer)) {
                  _context10.next = 6;
                  break;
                }

                return _context10.abrupt("return", this.createPayment(cart));

              case 6:
                if (!(cart.total && sessionData.amount === Math.round(cart.total))) {
                  _context10.next = 8;
                  break;
                }

                return _context10.abrupt("return", sessionData);

              case 8:
                return _context10.abrupt("return", this.stripe_.paymentIntents.update(sessionData.id, {
                  amount: Math.round(cart.total)
                }));

              case 9:
                _context10.next = 14;
                break;

              case 11:
                _context10.prev = 11;
                _context10.t0 = _context10["catch"](0);
                throw _context10.t0;

              case 14:
              case "end":
                return _context10.stop();
            }
          }
        }, _callee10, this, [[0, 11]]);
      }));

      function updatePayment(_x11, _x12) {
        return _updatePayment.apply(this, arguments);
      }

      return updatePayment;
    }()
  }, {
    key: "deletePayment",
    value: function () {
      var _deletePayment = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee11(payment) {
        var id;
        return regeneratorRuntime.wrap(function _callee11$(_context11) {
          while (1) {
            switch (_context11.prev = _context11.next) {
              case 0:
                _context11.prev = 0;
                id = payment.data.id;
                return _context11.abrupt("return", this.stripe_.paymentIntents.cancel(id)["catch"](function (err) {
                  if (err.statusCode === 400) {
                    return;
                  }

                  throw err;
                }));

              case 5:
                _context11.prev = 5;
                _context11.t0 = _context11["catch"](0);
                throw _context11.t0;

              case 8:
              case "end":
                return _context11.stop();
            }
          }
        }, _callee11, this, [[0, 5]]);
      }));

      function deletePayment(_x13) {
        return _deletePayment.apply(this, arguments);
      }

      return deletePayment;
    }()
    /**
     * Updates customer of Stripe payment intent.
     * @param {string} paymentIntentId - id of payment intent to update
     * @param {string} customerId - id of new Stripe customer
     * @returns {object} Stripe payment intent
     */

  }, {
    key: "updatePaymentIntentCustomer",
    value: function () {
      var _updatePaymentIntentCustomer = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee12(paymentIntentId, customerId) {
        return regeneratorRuntime.wrap(function _callee12$(_context12) {
          while (1) {
            switch (_context12.prev = _context12.next) {
              case 0:
                _context12.prev = 0;
                return _context12.abrupt("return", this.stripe_.paymentIntents.update(paymentIntentId, {
                  customer: customerId
                }));

              case 4:
                _context12.prev = 4;
                _context12.t0 = _context12["catch"](0);
                throw _context12.t0;

              case 7:
              case "end":
                return _context12.stop();
            }
          }
        }, _callee12, this, [[0, 4]]);
      }));

      function updatePaymentIntentCustomer(_x14, _x15) {
        return _updatePaymentIntentCustomer.apply(this, arguments);
      }

      return updatePaymentIntentCustomer;
    }()
    /**
     * Captures payment for Stripe payment intent.
     * @param {object} paymentData - payment method data from cart
     * @returns {object} Stripe payment intent
     */

  }, {
    key: "capturePayment",
    value: function () {
      var _capturePayment = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee13(payment) {
        var id, intent;
        return regeneratorRuntime.wrap(function _callee13$(_context13) {
          while (1) {
            switch (_context13.prev = _context13.next) {
              case 0:
                id = payment.data.id;
                _context13.prev = 1;
                _context13.next = 4;
                return this.stripe_.paymentIntents.capture(id);

              case 4:
                intent = _context13.sent;
                return _context13.abrupt("return", intent);

              case 8:
                _context13.prev = 8;
                _context13.t0 = _context13["catch"](1);

                if (!(_context13.t0.code === "payment_intent_unexpected_state")) {
                  _context13.next = 13;
                  break;
                }

                if (!(_context13.t0.payment_intent.status === "succeeded")) {
                  _context13.next = 13;
                  break;
                }

                return _context13.abrupt("return", _context13.t0.payment_intent);

              case 13:
                throw _context13.t0;

              case 14:
              case "end":
                return _context13.stop();
            }
          }
        }, _callee13, this, [[1, 8]]);
      }));

      function capturePayment(_x16) {
        return _capturePayment.apply(this, arguments);
      }

      return capturePayment;
    }()
    /**
     * Refunds payment for Stripe payment intent.
     * @param {object} paymentData - payment method data from cart
     * @param {number} amountToRefund - amount to refund
     * @returns {string} refunded payment intent
     */

  }, {
    key: "refundPayment",
    value: function () {
      var _refundPayment = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee14(payment, amountToRefund) {
        var id;
        return regeneratorRuntime.wrap(function _callee14$(_context14) {
          while (1) {
            switch (_context14.prev = _context14.next) {
              case 0:
                id = payment.data.id;
                _context14.prev = 1;
                _context14.next = 4;
                return this.stripe_.refunds.create({
                  amount: Math.round(amountToRefund),
                  payment_intent: id
                });

              case 4:
                return _context14.abrupt("return", payment.data);

              case 7:
                _context14.prev = 7;
                _context14.t0 = _context14["catch"](1);
                throw _context14.t0;

              case 10:
              case "end":
                return _context14.stop();
            }
          }
        }, _callee14, this, [[1, 7]]);
      }));

      function refundPayment(_x17, _x18) {
        return _refundPayment.apply(this, arguments);
      }

      return refundPayment;
    }()
    /**
     * Cancels payment for Stripe payment intent.
     * @param {object} paymentData - payment method data from cart
     * @returns {object} canceled payment intent
     */

  }, {
    key: "cancelPayment",
    value: function () {
      var _cancelPayment = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee15(payment) {
        var id;
        return regeneratorRuntime.wrap(function _callee15$(_context15) {
          while (1) {
            switch (_context15.prev = _context15.next) {
              case 0:
                id = payment.data.id;
                _context15.prev = 1;
                _context15.next = 4;
                return this.stripe_.paymentIntents.cancel(id);

              case 4:
                return _context15.abrupt("return", _context15.sent);

              case 7:
                _context15.prev = 7;
                _context15.t0 = _context15["catch"](1);

                if (!(_context15.t0.payment_intent.status === "canceled")) {
                  _context15.next = 11;
                  break;
                }

                return _context15.abrupt("return", _context15.t0.payment_intent);

              case 11:
                throw _context15.t0;

              case 12:
              case "end":
                return _context15.stop();
            }
          }
        }, _callee15, this, [[1, 7]]);
      }));

      function cancelPayment(_x19) {
        return _cancelPayment.apply(this, arguments);
      }

      return cancelPayment;
    }()
    /**
     * Constructs Stripe Webhook event
     * @param {object} data - the data of the webhook request: req.body
     * @param {object} signature - the Stripe signature on the event, that
     *    ensures integrity of the webhook event
     * @returns {object} Stripe Webhook event
     */

  }, {
    key: "constructWebhookEvent",
    value: function constructWebhookEvent(data, signature) {
      return this.stripe_.webhooks.constructEvent(data, signature, this.options_.webhook_secret);
    }
  }]);

  return StripeProviderService;
}(_medusaInterfaces.PaymentService);

_defineProperty(StripeProviderService, "identifier", "stripe");

var _default = StripeProviderService;
exports["default"] = _default;
