// medusa-extender/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Subscriber = exports.Validator = exports.Router = exports.Migration = exports.Middleware = exports.Repository = exports.Service = exports.Entity = exports.OnMedusaEntityEvent = exports.Module = exports.eventEmitter = exports.Medusa = exports.Utils = void 0;
__exportStar(require("./core/types"), exports);
__exportStar(require("./core/constants"), exports);
exports.Utils = require("./core/utils");
var Medusa_1 = require("./Medusa");
Object.defineProperty(exports, "Medusa", { enumerable: true, get: function () { return Medusa_1.Medusa; } });
var event_emmiter_1 = require("./core/event-emmiter");
Object.defineProperty(exports, "eventEmitter", { enumerable: true, get: function () { return event_emmiter_1.customEventEmitter; } });
var decorators_1 = require("./decorators");
Object.defineProperty(exports, "Module", { enumerable: true, get: function () { return decorators_1.Module; } });
Object.defineProperty(exports, "OnMedusaEntityEvent", { enumerable: true, get: function () { return decorators_1.OnMedusaEntityEvent; } });
Object.defineProperty(exports, "Entity", { enumerable: true, get: function () { return decorators_1.Entity; } });
Object.defineProperty(exports, "Service", { enumerable: true, get: function () { return decorators_1.Service; } });
Object.defineProperty(exports, "Repository", { enumerable: true, get: function () { return decorators_1.Repository; } });
Object.defineProperty(exports, "Middleware", { enumerable: true, get: function () { return decorators_1.Middleware; } });
Object.defineProperty(exports, "Migration", { enumerable: true, get: function () { return decorators_1.Migration; } });
Object.defineProperty(exports, "Router", { enumerable: true, get: function () { return decorators_1.Router; } });
Object.defineProperty(exports, "Validator", { enumerable: true, get: function () { return decorators_1.Validator; } });
Object.defineProperty(exports, "Subscriber", { enumerable: true, get: function () { return decorators_1.Subscriber; } });
__exportStar(require("./modules/multi-tenancy"), exports);
