// @medusajs/medusa/dist/api/routes/admin/orders/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.filterableAdminOrdersFields = exports.allowedAdminOrdersRelations = exports.allowedAdminOrdersFields = exports.defaultAdminOrdersFields = exports.defaultAdminOrdersRelations = void 0;
var express_1 = require("express");
require("reflect-metadata");
var common_1 = require("../../../../types/common");
var middlewares_1 = __importStar(require("../../../middlewares"));
var list_orders_1 = require("./list-orders");
var route = (0, express_1.Router)();
exports.default = (function (app, featureFlagRouter) {
    app.use("/orders", route);
    var relations = __spreadArray([], __read(exports.defaultAdminOrdersRelations), false);
    if (featureFlagRouter.isFeatureEnabled("sales_channels")) {
        relations.push("sales_channel");
    }
    /**
     * List orders
     */
    route.get("/", (0, middlewares_1.transformQuery)(list_orders_1.AdminGetOrdersParams, {
        defaultRelations: relations,
        defaultFields: exports.defaultAdminOrdersFields,
        allowedFields: exports.allowedAdminOrdersFields,
        isList: true,
    }), middlewares_1.default.wrap(require("./list-orders").default));
    /**
     * Get an order
     */
    route.get("/:id", (0, middlewares_1.transformQuery)(common_1.EmptyQueryParams, {
        defaultRelations: relations,
        defaultFields: exports.defaultAdminOrdersFields,
        allowedFields: exports.allowedAdminOrdersFields,
        isList: false,
    }), middlewares_1.default.wrap(require("./get-order").default));
    /**
     * Update an order
     */
    route.post("/:id", middlewares_1.default.wrap(require("./update-order").default));
    /**
     * Mark an order as completed
     */
    route.post("/:id/complete", middlewares_1.default.wrap(require("./complete-order").default));
    /**
     * Refund an amount to the customer's card.
     */
    route.post("/:id/refund", middlewares_1.default.wrap(require("./refund-payment").default));
    /**
     * Capture the authorized amount on the customer's card.
     */
    route.post("/:id/capture", middlewares_1.default.wrap(require("./capture-payment").default));
    /**
     * Create a fulfillment.
     */
    route.post("/:id/fulfillment", middlewares_1.default.wrap(require("./create-fulfillment").default));
    /**
     * Cancel a fulfillment related to an order.
     */
    route.post("/:id/fulfillments/:fulfillment_id/cancel", middlewares_1.default.wrap(require("./cancel-fulfillment").default));
    /**
     * Cancel a fulfillment related to a swap.
     */
    route.post("/:id/swaps/:swap_id/fulfillments/:fulfillment_id/cancel", middlewares_1.default.wrap(require("./cancel-fulfillment-swap").default));
    /**
     * Cancel a fulfillment related to a claim.
     */
    route.post("/:id/claims/:claim_id/fulfillments/:fulfillment_id/cancel", middlewares_1.default.wrap(require("./cancel-fulfillment-claim").default));
    /**
     * Create a shipment.
     */
    route.post("/:id/shipment", middlewares_1.default.wrap(require("./create-shipment").default));
    /**
     * Request a return.
     */
    route.post("/:id/return", middlewares_1.default.wrap(require("./request-return").default));
    /**
     * Cancel an order.
     */
    route.post("/:id/cancel", middlewares_1.default.wrap(require("./cancel-order").default));
    /**
     * Add a shipping method
     */
    route.post("/:id/shipping-methods", middlewares_1.default.wrap(require("./add-shipping-method").default));
    /**
     * Archive an order.
     */
    route.post("/:id/archive", middlewares_1.default.wrap(require("./archive-order").default));
    /**
     * Creates a swap, requests a return and prepares a cart for payment.
     */
    route.post("/:id/swaps", middlewares_1.default.wrap(require("./create-swap").default));
    /**
     * Cancels a swap.
     */
    route.post("/:id/swaps/:swap_id/cancel", middlewares_1.default.wrap(require("./cancel-swap").default));
    /**
     * Fulfills a swap.
     */
    route.post("/:id/swaps/:swap_id/fulfillments", middlewares_1.default.wrap(require("./fulfill-swap").default));
    /**
     * Marks a swap fulfillment as shipped.
     */
    route.post("/:id/swaps/:swap_id/shipments", middlewares_1.default.wrap(require("./create-swap-shipment").default));
    /**
     * Captures the payment associated with a swap
     */
    route.post("/:id/swaps/:swap_id/process-payment", middlewares_1.default.wrap(require("./process-swap-payment").default));
    /**
     * Creates a claim
     */
    route.post("/:id/claims", middlewares_1.default.wrap(require("./create-claim").default));
    /**
     * Cancels a claim
     */
    route.post("/:id/claims/:claim_id/cancel", middlewares_1.default.wrap(require("./cancel-claim").default));
    /**
     * Updates a claim
     */
    route.post("/:id/claims/:claim_id", middlewares_1.default.wrap(require("./update-claim").default));
    /**
     * Creates claim fulfillment
     */
    route.post("/:id/claims/:claim_id/fulfillments", middlewares_1.default.wrap(require("./fulfill-claim").default));
    /**
     * Creates claim shipment
     */
    route.post("/:id/claims/:claim_id/shipments", middlewares_1.default.wrap(require("./create-claim-shipment").default));
    return app;
});
exports.defaultAdminOrdersRelations = [
    "customer",
    "billing_address",
    "shipping_address",
    "discounts",
    "discounts.rule",
    "shipping_methods",
    "payments",
    "fulfillments",
    "fulfillments.tracking_links",
    "fulfillments.items",
    "returns",
    "returns.shipping_method",
    "returns.shipping_method.tax_lines",
    "returns.items",
    "returns.items.reason",
    "gift_cards",
    "gift_card_transactions",
    "claims",
    "claims.return_order",
    "claims.return_order.shipping_method",
    "claims.shipping_methods",
    "claims.shipping_address",
    "claims.additional_items",
    "claims.fulfillments",
    "claims.fulfillments.tracking_links",
    "claims.claim_items",
    "claims.claim_items.item",
    "claims.claim_items.images",
    // "claims.claim_items.tags",
    "swaps",
    "swaps.return_order",
    "swaps.payment",
    "swaps.shipping_methods",
    "swaps.shipping_address",
    "swaps.additional_items",
    "swaps.fulfillments",
    "swaps.fulfillments.tracking_links",
];
exports.defaultAdminOrdersFields = [
    "id",
    "status",
    "fulfillment_status",
    "payment_status",
    "display_id",
    "cart_id",
    "draft_order_id",
    "customer_id",
    "email",
    "region_id",
    "currency_code",
    "tax_rate",
    "canceled_at",
    "created_at",
    "updated_at",
    "metadata",
    "items.refundable",
    "swaps.additional_items.refundable",
    "claims.additional_items.refundable",
    "shipping_total",
    "discount_total",
    "tax_total",
    "refunded_total",
    "gift_card_total",
    "subtotal",
    "total",
    "paid_total",
    "refundable_amount",
    "no_notification",
];
exports.allowedAdminOrdersFields = [
    "id",
    "status",
    "fulfillment_status",
    "payment_status",
    "display_id",
    "cart_id",
    "draft_order_id",
    "customer_id",
    "email",
    "region_id",
    "currency_code",
    "tax_rate",
    "canceled_at",
    "created_at",
    "updated_at",
    "metadata",
    "shipping_total",
    "discount_total",
    "tax_total",
    "refunded_total",
    "subtotal",
    "gift_card_total",
    "total",
    "paid_total",
    "refundable_amount",
    "no_notification",
];
exports.allowedAdminOrdersRelations = [
    "customer",
    "region",
    "billing_address",
    "shipping_address",
    "discounts",
    "discounts.rule",
    "shipping_methods",
    "payments",
    "fulfillments",
    "fulfillments.tracking_links",
    "returns",
    "claims",
    "swaps",
    "swaps.return_order",
    "swaps.additional_items",
];
exports.filterableAdminOrdersFields = [
    "id",
    "status",
    "fulfillment_status",
    "payment_status",
    "display_id",
    "cart_id",
    "customer_id",
    "email",
    "region_id",
    "currency_code",
    "tax_rate",
    "canceled_at",
    "created_at",
    "updated_at",
];
__exportStar(require("./add-shipping-method"), exports);
__exportStar(require("./archive-order"), exports);
__exportStar(require("./cancel-claim"), exports);
__exportStar(require("./cancel-fulfillment"), exports);
__exportStar(require("./cancel-fulfillment-claim"), exports);
__exportStar(require("./cancel-fulfillment-swap"), exports);
__exportStar(require("./cancel-order"), exports);
__exportStar(require("./cancel-swap"), exports);
__exportStar(require("./capture-payment"), exports);
__exportStar(require("./complete-order"), exports);
__exportStar(require("./create-claim"), exports);
__exportStar(require("./create-claim-shipment"), exports);
__exportStar(require("./create-fulfillment"), exports);
__exportStar(require("./create-shipment"), exports);
__exportStar(require("./create-swap"), exports);
__exportStar(require("./create-swap-shipment"), exports);
__exportStar(require("./fulfill-claim"), exports);
__exportStar(require("./fulfill-swap"), exports);
__exportStar(require("./get-order"), exports);
__exportStar(require("./list-orders"), exports);
__exportStar(require("./process-swap-payment"), exports);
__exportStar(require("./refund-payment"), exports);
__exportStar(require("./request-return"), exports);
__exportStar(require("./update-claim"), exports);
__exportStar(require("./update-order"), exports);
