// medusa-extender/dist/core/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MEDUSA_RESOLVER_KEYS = exports.INJECTABLE_OPTIONS_KEY = exports.MODULE_KEY = void 0;
/**
 * Metadata key to store config from decorated injectables.
 */
exports.MODULE_KEY = '__MODULE_KEY__';
/**
 * Metadata key to store config from decorated injectables.
 */
exports.INJECTABLE_OPTIONS_KEY = '__INJECTABLE_OPTIONS_KEY__';
/**
 * Hold the name of the medusa services registered into the container for later on resolve.
 */
exports.MEDUSA_RESOLVER_KEYS = {
    EventBusService: 'eventbusService',
    UserService: 'userService',
    CustomerService: 'customerService',
    AuthService: 'authService',
    MiddlewareService: 'middlewareService',
    NotificationService: 'notificationService',
    TotalsService: 'totalsService',
    StoreService: 'storeService',
    PaymentProviderService: 'paymentproviderService',
    FulfillmentProviderService: 'fulfillmentproviderService',
    RegionService: 'regionService',
    ProductVariantService: 'productvariantService',
    ProductCollectionService: 'productcollectionService',
    DefaultSearchService: 'defaultsearchService',
    ProductService: 'productService',
    DiscountService: 'discountService',
    GiftCardService: 'giftcardService',
    ShippingOptionService: 'shippingoptionService',
    CustomShippingOptionService: 'customshippingoptionService',
    ShippingProfileService: 'shippingprofileService',
    LineItemService: 'lineitemService',
    FulfillmentService: 'fulfillmentService',
    InventoryService: 'inventoryService',
    CartService: 'cartService',
    DraftOrderService: 'draftOrderService',
    OrderService: 'orderService',
    manager: 'manager',
    logger: 'logger',
    requestContext: 'requestContext',
};
