// @medusajs/medusa/dist/services/payment-provider.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var medusa_core_utils_1 = require("medusa-core-utils");
var interfaces_1 = require("../interfaces");
var utils_1 = require("../utils");
/**
 * Helps retrieve payment providers
 */
var PaymentProviderService = /** @class */ (function (_super) {
    __extends(PaymentProviderService, _super);
    function PaymentProviderService(container) {
        var _this = _super.call(this, container) || this;
        _this.container_ = container;
        _this.manager_ = container.manager;
        _this.paymentSessionRepository_ = container.paymentSessionRepository;
        _this.paymentProviderRepository_ = container.paymentProviderRepository;
        _this.paymentRepository_ = container.paymentRepository;
        _this.refundRepository_ = container.refundRepository;
        return _this;
    }
    PaymentProviderService.prototype.registerInstalledProviders = function (providerIds) {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.atomicPhase_(function (transactionManager) { return __awaiter(_this, void 0, void 0, function () {
                            var model;
                            var _this = this;
                            return __generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0:
                                        model = transactionManager.getCustomRepository(this.paymentProviderRepository_);
                                        return [4 /*yield*/, model.update({}, { is_installed: false })];
                                    case 1:
                                        _a.sent();
                                        return [4 /*yield*/, Promise.all(providerIds.map(function (providerId) { return __awaiter(_this, void 0, void 0, function () {
                                                var provider;
                                                return __generator(this, function (_a) {
                                                    switch (_a.label) {
                                                        case 0:
                                                            provider = model.create({
                                                                id: providerId,
                                                                is_installed: true,
                                                            });
                                                            return [4 /*yield*/, model.save(provider)];
                                                        case 1: return [2 /*return*/, _a.sent()];
                                                    }
                                                });
                                            }); }))];
                                    case 2:
                                        _a.sent();
                                        return [2 /*return*/];
                                }
                            });
                        }); })];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    PaymentProviderService.prototype.list = function () {
        return __awaiter(this, void 0, void 0, function () {
            var ppRepo;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        ppRepo = this.manager_.getCustomRepository(this.paymentProviderRepository_);
                        return [4 /*yield*/, ppRepo.find()];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    PaymentProviderService.prototype.retrievePayment = function (id, relations) {
        if (relations === void 0) { relations = []; }
        return __awaiter(this, void 0, void 0, function () {
            var paymentRepo, query, payment;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        paymentRepo = this.manager_.getCustomRepository(this.paymentRepository_);
                        query = {
                            where: { id: id },
                            relations: [],
                        };
                        if (relations.length) {
                            query.relations = relations;
                        }
                        return [4 /*yield*/, paymentRepo.findOne(query)];
                    case 1:
                        payment = _a.sent();
                        if (!payment) {
                            throw new medusa_core_utils_1.MedusaError(medusa_core_utils_1.MedusaError.Types.NOT_FOUND, "Payment with ".concat(id, " was not found"));
                        }
                        return [2 /*return*/, payment];
                }
            });
        });
    };
    PaymentProviderService.prototype.listPayments = function (selector, config) {
        if (config === void 0) { config = {
            skip: 0,
            take: 50,
            order: { created_at: "DESC" },
        }; }
        return __awaiter(this, void 0, void 0, function () {
            var payRepo, query;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        payRepo = this.manager_.getCustomRepository(this.paymentRepository_);
                        query = (0, utils_1.buildQuery)(selector, config);
                        return [4 /*yield*/, payRepo.find(query)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    PaymentProviderService.prototype.retrieveSession = function (id, relations) {
        if (relations === void 0) { relations = []; }
        return __awaiter(this, void 0, void 0, function () {
            var sessionRepo, query, session;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        sessionRepo = this.manager_.getCustomRepository(this.paymentSessionRepository_);
                        query = {
                            where: { id: id },
                            relations: [],
                        };
                        if (relations.length) {
                            query.relations = relations;
                        }
                        return [4 /*yield*/, sessionRepo.findOne(query)];
                    case 1:
                        session = _a.sent();
                        if (!session) {
                            throw new medusa_core_utils_1.MedusaError(medusa_core_utils_1.MedusaError.Types.NOT_FOUND, "Payment Session with ".concat(id, " was not found"));
                        }
                        return [2 /*return*/, session];
                }
            });
        });
    };
    /**
     * Creates a payment session with the given provider.
     * @param providerId - the id of the provider to create payment with
     * @param cart - a cart object used to calculate the amount, etc. from
     * @return the payment session
     */
    PaymentProviderService.prototype.createSession = function (providerId, cart) {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.atomicPhase_(function (transactionManager) { return __awaiter(_this, void 0, void 0, function () {
                            var provider, sessionData, sessionRepo, toCreate, created;
                            return __generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0:
                                        provider = this.retrieveProvider(providerId);
                                        return [4 /*yield*/, provider
                                                .withTransaction(transactionManager)
                                                .createPayment(cart)];
                                    case 1:
                                        sessionData = _a.sent();
                                        sessionRepo = transactionManager.getCustomRepository(this.paymentSessionRepository_);
                                        toCreate = {
                                            cart_id: cart.id,
                                            provider_id: providerId,
                                            data: sessionData,
                                            status: "pending",
                                        };
                                        created = sessionRepo.create(toCreate);
                                        return [4 /*yield*/, sessionRepo.save(created)];
                                    case 2: return [2 /*return*/, _a.sent()];
                                }
                            });
                        }); })];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    /**
     * Refreshes a payment session with the given provider.
     * This means, that we delete the current one and create a new.
     * @param paymentSession - the payment session object to
     *    update
     * @param cart - a cart object used to calculate the amount, etc. from
     * @return the payment session
     */
    PaymentProviderService.prototype.refreshSession = function (paymentSession, cart) {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                return [2 /*return*/, this.atomicPhase_(function (transactionManager) { return __awaiter(_this, void 0, void 0, function () {
                        var session, provider, sessionRepo, sessionData, toCreate, created;
                        return __generator(this, function (_a) {
                            switch (_a.label) {
                                case 0: return [4 /*yield*/, this.retrieveSession(paymentSession.id)];
                                case 1:
                                    session = _a.sent();
                                    provider = this.retrieveProvider(paymentSession.provider_id);
                                    return [4 /*yield*/, provider.withTransaction(transactionManager).deletePayment(session)];
                                case 2:
                                    _a.sent();
                                    sessionRepo = transactionManager.getCustomRepository(this.paymentSessionRepository_);
                                    return [4 /*yield*/, sessionRepo.remove(session)];
                                case 3:
                                    _a.sent();
                                    return [4 /*yield*/, provider
                                            .withTransaction(transactionManager)
                                            .createPayment(cart)];
                                case 4:
                                    sessionData = _a.sent();
                                    toCreate = {
                                        cart_id: cart.id,
                                        provider_id: session.provider_id,
                                        data: sessionData,
                                        is_selected: true,
                                        status: "pending",
                                    };
                                    created = sessionRepo.create(toCreate);
                                    return [4 /*yield*/, sessionRepo.save(created)];
                                case 5: return [2 /*return*/, _a.sent()];
                            }
                        });
                    }); })];
            });
        });
    };
    /**
     * Updates an existing payment session.
     * @param paymentSession - the payment session object to
     *    update
     * @param cart - the cart object to update for
     * @return the updated payment session
     */
    PaymentProviderService.prototype.updateSession = function (paymentSession, cart) {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.atomicPhase_(function (transactionManager) { return __awaiter(_this, void 0, void 0, function () {
                            var session, provider, _a, sessionRepo;
                            return __generator(this, function (_b) {
                                switch (_b.label) {
                                    case 0: return [4 /*yield*/, this.retrieveSession(paymentSession.id)];
                                    case 1:
                                        session = _b.sent();
                                        provider = this.retrieveProvider(paymentSession.provider_id);
                                        _a = session;
                                        return [4 /*yield*/, provider
                                                .withTransaction(transactionManager)
                                                .updatePayment(paymentSession.data, cart)];
                                    case 2:
                                        _a.data = _b.sent();
                                        sessionRepo = transactionManager.getCustomRepository(this.paymentSessionRepository_);
                                        return [2 /*return*/, sessionRepo.save(session)];
                                }
                            });
                        }); })];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    PaymentProviderService.prototype.deleteSession = function (paymentSession) {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.atomicPhase_(function (transactionManager) { return __awaiter(_this, void 0, void 0, function () {
                            var session, provider, sessionRepo;
                            return __generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0: return [4 /*yield*/, this.retrieveSession(paymentSession.id).catch(function () { return void 0; })];
                                    case 1:
                                        session = _a.sent();
                                        if (!session) {
                                            return [2 /*return*/];
                                        }
                                        provider = this.retrieveProvider(paymentSession.provider_id);
                                        return [4 /*yield*/, provider
                                                .withTransaction(transactionManager)
                                                .deletePayment(paymentSession)];
                                    case 2:
                                        _a.sent();
                                        sessionRepo = transactionManager.getCustomRepository(this.paymentSessionRepository_);
                                        return [2 /*return*/, sessionRepo.remove(session)];
                                }
                            });
                        }); })];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    /**
     * Finds a provider given an id
     * @param {string} providerId - the id of the provider to get
     * @return {PaymentService} the payment provider
     */
    PaymentProviderService.prototype.retrieveProvider = function (providerId) {
        try {
            var provider = void 0;
            if (providerId === "system") {
                provider = this.container_["systemPaymentProviderService"];
            }
            else {
                provider = this.container_["pp_".concat(providerId)];
            }
            return provider;
        }
        catch (err) {
            throw new medusa_core_utils_1.MedusaError(medusa_core_utils_1.MedusaError.Types.NOT_FOUND, "Could not find a payment provider with id: ".concat(providerId));
        }
    };
    PaymentProviderService.prototype.createPayment = function (cart) {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.atomicPhase_(function (transactionManager) { return __awaiter(_this, void 0, void 0, function () {
                            var paymentSession, region, total, provider, paymentData, paymentRepo, created;
                            return __generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0:
                                        paymentSession = cart.payment_session, region = cart.region, total = cart.total;
                                        provider = this.retrieveProvider(paymentSession.provider_id);
                                        return [4 /*yield*/, provider
                                                .withTransaction(transactionManager)
                                                .getPaymentData(paymentSession)];
                                    case 1:
                                        paymentData = _a.sent();
                                        paymentRepo = transactionManager.getCustomRepository(this.paymentRepository_);
                                        created = paymentRepo.create({
                                            provider_id: paymentSession.provider_id,
                                            amount: total,
                                            currency_code: region.currency_code,
                                            data: paymentData,
                                            cart_id: cart.id,
                                        });
                                        return [2 /*return*/, paymentRepo.save(created)];
                                }
                            });
                        }); })];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    PaymentProviderService.prototype.updatePayment = function (paymentId, data) {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.atomicPhase_(function (transactionManager) { return __awaiter(_this, void 0, void 0, function () {
                            var payment, payRepo;
                            return __generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0: return [4 /*yield*/, this.retrievePayment(paymentId)];
                                    case 1:
                                        payment = _a.sent();
                                        if (data === null || data === void 0 ? void 0 : data.order_id) {
                                            payment.order_id = data.order_id;
                                        }
                                        if (data === null || data === void 0 ? void 0 : data.swap_id) {
                                            payment.swap_id = data.swap_id;
                                        }
                                        payRepo = transactionManager.getCustomRepository(this.paymentRepository_);
                                        return [2 /*return*/, payRepo.save(payment)];
                                }
                            });
                        }); })];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    PaymentProviderService.prototype.authorizePayment = function (paymentSession, context) {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.atomicPhase_(function (transactionManager) { return __awaiter(_this, void 0, void 0, function () {
                            var session, provider, _a, status, data, sessionRepo;
                            return __generator(this, function (_b) {
                                switch (_b.label) {
                                    case 0: return [4 /*yield*/, this.retrieveSession(paymentSession.id).catch(function () { return void 0; })];
                                    case 1:
                                        session = _b.sent();
                                        if (!session) {
                                            return [2 /*return*/];
                                        }
                                        provider = this.retrieveProvider(paymentSession.provider_id);
                                        return [4 /*yield*/, provider
                                                .withTransaction(transactionManager)
                                                .authorizePayment(session, context)];
                                    case 2:
                                        _a = _b.sent(), status = _a.status, data = _a.data;
                                        session.data = data;
                                        session.status = status;
                                        sessionRepo = transactionManager.getCustomRepository(this.paymentSessionRepository_);
                                        return [2 /*return*/, sessionRepo.save(session)];
                                }
                            });
                        }); })];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    PaymentProviderService.prototype.updateSessionData = function (paymentSession, data) {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.atomicPhase_(function (transactionManager) { return __awaiter(_this, void 0, void 0, function () {
                            var session, provider, _a, sessionRepo;
                            return __generator(this, function (_b) {
                                switch (_b.label) {
                                    case 0: return [4 /*yield*/, this.retrieveSession(paymentSession.id)];
                                    case 1:
                                        session = _b.sent();
                                        provider = this.retrieveProvider(paymentSession.provider_id);
                                        _a = session;
                                        return [4 /*yield*/, provider
                                                .withTransaction(transactionManager)
                                                .updatePaymentData(paymentSession.data, data)];
                                    case 2:
                                        _a.data = _b.sent();
                                        session.status = paymentSession.status;
                                        sessionRepo = transactionManager.getCustomRepository(this.paymentSessionRepository_);
                                        return [2 /*return*/, sessionRepo.save(session)];
                                }
                            });
                        }); })];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    PaymentProviderService.prototype.cancelPayment = function (paymentObj) {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.atomicPhase_(function (transactionManager) { return __awaiter(_this, void 0, void 0, function () {
                            var payment, provider, _a, now, paymentRepo;
                            return __generator(this, function (_b) {
                                switch (_b.label) {
                                    case 0: return [4 /*yield*/, this.retrievePayment(paymentObj.id)];
                                    case 1:
                                        payment = _b.sent();
                                        provider = this.retrieveProvider(payment.provider_id);
                                        _a = payment;
                                        return [4 /*yield*/, provider
                                                .withTransaction(transactionManager)
                                                .cancelPayment(payment)];
                                    case 2:
                                        _a.data = _b.sent();
                                        now = new Date();
                                        payment.canceled_at = now.toISOString();
                                        paymentRepo = transactionManager.getCustomRepository(this.paymentRepository_);
                                        return [4 /*yield*/, paymentRepo.save(payment)];
                                    case 3: return [2 /*return*/, _b.sent()];
                                }
                            });
                        }); })];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    PaymentProviderService.prototype.getStatus = function (payment) {
        return __awaiter(this, void 0, void 0, function () {
            var provider;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        provider = this.retrieveProvider(payment.provider_id);
                        return [4 /*yield*/, provider.withTransaction(this.manager_).getStatus(payment.data)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    PaymentProviderService.prototype.capturePayment = function (paymentObj) {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.atomicPhase_(function (transactionManager) { return __awaiter(_this, void 0, void 0, function () {
                            var payment, provider, _a, now, paymentRepo;
                            return __generator(this, function (_b) {
                                switch (_b.label) {
                                    case 0: return [4 /*yield*/, this.retrievePayment(paymentObj.id)];
                                    case 1:
                                        payment = _b.sent();
                                        provider = this.retrieveProvider(payment.provider_id);
                                        _a = payment;
                                        return [4 /*yield*/, provider
                                                .withTransaction(transactionManager)
                                                .capturePayment(payment)];
                                    case 2:
                                        _a.data = _b.sent();
                                        now = new Date();
                                        payment.captured_at = now.toISOString();
                                        paymentRepo = transactionManager.getCustomRepository(this.paymentRepository_);
                                        return [2 /*return*/, paymentRepo.save(payment)];
                                }
                            });
                        }); })];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    PaymentProviderService.prototype.refundPayment = function (payObjs, amount, reason, note) {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.atomicPhase_(function (transactionManager) { return __awaiter(_this, void 0, void 0, function () {
                            var payments, order_id, refundable, balance, used, paymentRepo, paymentToRefund, currentRefundable, refundAmount, provider, _a, refundRepo, toCreate, created;
                            return __generator(this, function (_b) {
                                switch (_b.label) {
                                    case 0: return [4 /*yield*/, this.listPayments({
                                            id: payObjs.map(function (p) { return p.id; }),
                                        })];
                                    case 1:
                                        payments = _b.sent();
                                        refundable = payments.reduce(function (acc, next) {
                                            order_id = next.order_id;
                                            if (next.captured_at) {
                                                return (acc += next.amount - next.amount_refunded);
                                            }
                                            return acc;
                                        }, 0);
                                        if (refundable < amount) {
                                            throw new medusa_core_utils_1.MedusaError(medusa_core_utils_1.MedusaError.Types.NOT_ALLOWED, "Refund amount is higher that the refundable amount");
                                        }
                                        balance = amount;
                                        used = [];
                                        paymentRepo = transactionManager.getCustomRepository(this.paymentRepository_);
                                        paymentToRefund = payments.find(function (payment) { return payment.amount - payment.amount_refunded > 0; });
                                        _b.label = 2;
                                    case 2:
                                        if (!paymentToRefund) return [3 /*break*/, 5];
                                        currentRefundable = paymentToRefund.amount - paymentToRefund.amount_refunded;
                                        refundAmount = Math.min(currentRefundable, balance);
                                        provider = this.retrieveProvider(paymentToRefund.provider_id);
                                        _a = paymentToRefund;
                                        return [4 /*yield*/, provider
                                                .withTransaction(transactionManager)
                                                .refundPayment(paymentToRefund, refundAmount)];
                                    case 3:
                                        _a.data = _b.sent();
                                        paymentToRefund.amount_refunded += refundAmount;
                                        return [4 /*yield*/, paymentRepo.save(paymentToRefund)];
                                    case 4:
                                        _b.sent();
                                        balance -= refundAmount;
                                        used.push(paymentToRefund.id);
                                        if (balance > 0) {
                                            paymentToRefund = payments.find(function (payment) {
                                                return payment.amount - payment.amount_refunded > 0 &&
                                                    !used.includes(payment.id);
                                            });
                                        }
                                        else {
                                            paymentToRefund = undefined;
                                        }
                                        return [3 /*break*/, 2];
                                    case 5:
                                        refundRepo = transactionManager.getCustomRepository(this.refundRepository_);
                                        toCreate = {
                                            order_id: order_id,
                                            amount: amount,
                                            reason: reason,
                                            note: note,
                                        };
                                        created = refundRepo.create(toCreate);
                                        return [2 /*return*/, refundRepo.save(created)];
                                }
                            });
                        }); })];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    PaymentProviderService.prototype.retrieveRefund = function (id, config) {
        if (config === void 0) { config = {}; }
        return __awaiter(this, void 0, void 0, function () {
            var refRepo, query, refund;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        refRepo = this.manager_.getCustomRepository(this.refundRepository_);
                        query = (0, utils_1.buildQuery)({ id: id }, config);
                        return [4 /*yield*/, refRepo.findOne(query)];
                    case 1:
                        refund = _a.sent();
                        if (!refund) {
                            throw new medusa_core_utils_1.MedusaError(medusa_core_utils_1.MedusaError.Types.NOT_FOUND, "A refund with ".concat(id, " was not found"));
                        }
                        return [2 /*return*/, refund];
                }
            });
        });
    };
    return PaymentProviderService;
}(interfaces_1.TransactionBaseService));
exports.default = PaymentProviderService;
